tests/mapper/test_graphwave_generator.py

```diff
@@ -0,0 +1,272 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2020 Data61, CSIRO
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from stellargraph.core.graph import *
+from stellargraph.mapper.graphwave_generator import (
+    GraphWaveGenerator,
+    _empirical_characteristic_function,
+)
+
+import networkx as nx
+import numpy as np
+import pytest
+import scipy.sparse as sps
+import tensorflow as tf
+
+
+class TestGraphWave:
+
+    gnx = nx.barbell_graph(m1=10, m2=11)
```

Comment:
Putting initialisation code in the class instance forces it to run as the file is being loaded, rather than only when the tests are being run. This was the problem in https://github.com/stellargraph/stellargraph/pull/791. Maybe this moderately complicated initialisation could use fixtures instead? https://docs.pytest.org/en/latest/fixture.html

```python
@pytest.fixture
def barbell():
    return StellarGraph.from_networkx(nx.barbell_graph(m1=10, m2=11))

...

def test_init(self, barbell):
    generator = GraphWaveGenerator(barbell, ...)
```

I think `num_nodes` probably doesn't need to be cached: it's `G.number_of_nodes()` or `barbell.number_of_nodes()`, and `sample_points` is cheap enough that it can stay as globally computed.